RetailCoder.VBE/Refactorings/Rename/RenameRefactoring.cs

```diff
@@ -38,358 +40,496 @@ public void Refactor()
+
+                RenameDeclaration(_model, eventDeclaration, newEventName);
+                RewriteContent(_model);
+
+                _model.State.OnParseRequested(this);
             }
         }
 
-        private void RenameControl()
+        internal class RenameControlHandler : IRenameRefactoringHandler
         {
-            try
+            private Declaration _target;
+            private RenameModel _model;
+
+            public RenameControlHandler(RenameModel model)
+            {
+                _target = model.Target;
+                _model = model;
+            }
+
+            //todo: Add error message string to UI Resources
+            public string ErrorMessage { get { return "RenameDialog_ControlRenameError"; } }
+
+            public void Rename()
             {
+                var controlDeclaration = GetDeclarationOfType(_model, DeclarationType.Control, false);
+                if (null == controlDeclaration) { return; }
+
                 var module = _model.Target.QualifiedName.QualifiedModuleName.Component.CodeModule;
                 var component = module.Parent;
-                var control = component.Controls.SingleOrDefault(item => item.Name == _model.Target.IdentifierName);
+                var control = component.Controls.SingleOrDefault(item => item.Name == controlDeclaration.IdentifierName);
+
+                if (control == null) { return; }
+
+                var newControlName = _model.NewName;
+                if (newControlName.Contains("_"))
                 {
-                    if (control == null)
-                    {
-                        return;
-                    }
+                    newControlName = _model.NewName.Remove(_model.NewName.IndexOf("_"));
+                }
 
-                    foreach (var handler in _model.Declarations.FindEventHandlers(_model.Target).OrderByDescending(h => h.Selection.StartColumn))
-                    {
-                        var newMemberName = handler.IdentifierName.Replace(control.Name + '_', _model.NewName + '_');
-                        var project = handler.Project;
-                        var components = project.VBComponents;
-                        var refComponent = components[handler.ComponentName];
-                        var refModule = refComponent.CodeModule;
-                        {
-                            var content = refModule.GetLines(handler.Selection.StartLine, 1);
-                            var newContent = GetReplacementLine(content, newMemberName, handler.Selection);
-                            refModule.ReplaceLine(handler.Selection.StartLine, newContent);
-                        }
-                    }
+                //User attempted to rename an event sub signature e.g. btn1_Click() to btn1_ClickAgain()
+                if (newControlName.Equals(controlDeclaration.IdentifierName))
+                {
+                    //nothing to change
+                    return;
```

Comment:
@BZngr absolutely - that's actually an oversight: if we're renaming a method, and that method is implementing an interface member or handling an event, then we need to set the refactoring target to the interface member or whatever object is exposing that event - i.e. we're renaming a control, not a procedure. Good catch! :+1: